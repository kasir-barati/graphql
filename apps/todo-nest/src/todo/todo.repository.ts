import { Injectable } from '@nestjs/common';
import { PrismaService } from '@shared';

@Injectable()
export class TodoRepository {
  constructor(private readonly prismaService: PrismaService) {}

  read(id: string) {
    return this.prismaService.todo.findFirst({
      where: { id },
      include: {
        AssignedTo: true,
        CreatedBy: true,
      },
    });
  }

  create({
    title,
    content,
    createdById,
    assignedToId,
  }: {
    title: string;
    content: string;
    createdById: string;
    assignedToId?: string;
  }) {
    return this.prismaService.todo.create({
      data: {
        title,
        content,
        createdById,
        assignedToId,
      },
      select: {
        id: true,
        title: true,
        content: true,
        createdAt: true,
        updatedAt: true,
        CreatedBy: true,
        createdById: true,
        ...(assignedToId && { AssignedTo: true, assignedToId: true }),
      },
    });
  }

  update(
    id: string,
    {
      title,
      content,
      createdById,
      assignedToId,
    }: {
      title?: string;
      content?: string;
      createdById?: string;
      assignedToId?: string;
    },
  ) {
    return this.prismaService.todo.update({
      where: {
        id,
      },
      data: {
        title,
        content,
        createdById,
        assignedToId,
      },
      include: {
        CreatedBy: true,
        AssignedTo: true,
      },
    });
  }

  async delete(id: string) {
    await this.prismaService.todo.delete({
      where: {
        id,
      },
    });
  }
}
