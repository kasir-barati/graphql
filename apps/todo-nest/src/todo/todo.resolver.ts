import { Args, ID, Query, Resolver } from '@nestjs/graphql';
import { UserService } from '../user/user.service';
import { Todo } from './models/todo.model';
import { TodoService } from './todo.service';

@Resolver(() => Todo)
export class TodoResolver {
  constructor(
    private readonly todoService: TodoService,
    private readonly userService: UserService,
  ) {}

  @Query(() => Todo, { nullable: true })
  getTodo(
    @Args('id', { type: () => ID, name: 'getTodo' }) id: string,
  ) {
    return this.todoService.getTodo(id);
  }

  // Learn more about ResolveField: https://github.com/kasir-barati/graphql/blob/main/docs/nestjs.md#shouldWeUseResolveField
  // @ResolveField(() => User, { nullable: true, name: 'AssignedTo' })
  // async getAssignedTo(@Parent() todo: Todo) {
  //   // Since this field is nullable we need to check if that field exits
  //   if (!todo.assignedToId) {
  //     return;
  //   }

  //   return this.userService.getUser(todo.assignedToId);
  // }

  // @ResolveField(() => User, { nullable: true, name: 'CreatedBy' })
  // async getCreatedBy(@Parent() todo: Todo) {
  //   return this.userService.getUser(todo.createdById);
  // }
}
