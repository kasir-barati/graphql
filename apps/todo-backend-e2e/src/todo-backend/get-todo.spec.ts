import axios from 'axios';
import { GetTodoResponse } from '../support/types/todo.type';

describe('POST /graphql', () => {
  it('should get a todo', async () => {
    const query = /* GraphQL */ `
      {
        getTodo(id: "eec5e3aa-7137-4c9c-a723-ec2f43d4daa4") {
          id
          content
        }
      }
    `;

    const {
      data: { data },
      status,
    } = await axios.post<GetTodoResponse>(`/graphql`, {
      query,
    });

    expect(status).toBe(200);
    expect(data).toMatchSnapshot();
  });
});
