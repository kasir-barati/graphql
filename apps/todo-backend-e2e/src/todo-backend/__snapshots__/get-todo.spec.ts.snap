// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /graphql should get a todo 1`] = `
{
  "getTodo": {
    "content": "Some content",
    "id": "eec5e3aa-7137-4c9c-a723-ec2f43d4daa4",
  },
}
`;

exports[`POST /graphql should get a todo and rename fields to snake_case 1`] = `
{
  "getTodo": {
    "content": "Some content",
    "created_by": {
      "username": "kasir-barati",
    },
    "id": "eec5e3aa-7137-4c9c-a723-ec2f43d4daa4",
  },
}
`;
